Append ? on types annotated with nullable attributes

MaStr11 opened the pull request:
Fixes #48023

MaStr11 commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -143,7 +143,10 @@ public override void VisitProperty(IPropertySymbol symbol)
+                var symbolType = ShouldBeNullableAnnotated(symbol, FlowAnalysisAnnotations.MaybeNull)
+                    ? symbol.Type.WithNullableAnnotation(CodeAnalysis.NullableAnnotation.Annotated)
+                    : symbol.Type;
+                symbolType.Accept(this.NotFirstVisitor);
```
This is one possible solution to append the question mark. I don't know whether that's the right approach for the problem.

MaStr11 reviewed (commented):
The current draft shows one possible solution: Fix the `SymbolDisplayVistor` to append the ? to the types based on `GetFlowAnalysisAnnotations`. If this is the right approach, I would continue with other nullable attributes and locations (e.g. parameters).

MaStr11 commented:
# Spec of when to show a question mark

Under which circumstances should a question mark on elements marked with a [reserved nullable attribute](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis) be shown.

### AllowNull

`AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter`

Probably not for property and field. I'm not sure for `out` and `ref` parameters.

### DisallowNull

Irrelevant.

### MaybeNull
`AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter | AttributeTargets.ReturnValue`

Always.

### NotNull

Irrelevant.

### MaybeNullWhen
`AttributeTargets.Parameter`

Always.

### NotNullWhen

Irrelevant.

### NotNullIfNotNull

Irrelevant.

### DoesNotReturn

Irrelevant.

### DoesNotReturnIf

Irrelevant.

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -143,7 +143,10 @@ public override void VisitProperty(IPropertySymbol symbol)
                 AddCustomModifiersIfRequired(symbol.RefCustomModifiers);
 
-                symbol.Type.Accept(this.NotFirstVisitor);
+                var symbolType = ShouldBeNullableAnnotated(symbol, FlowAnalysisAnnotations.MaybeNull)
```
>ShouldBeNullableAnnotated(symbol, FlowAnalysisAnnotations.MaybeNull) [](start = 33, length = 68)

It doesn't feel right to call this when ```symbol.Type``` is a value type.

MaStr11 replied:
Fixed in 57b130db

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -172,6 +175,18 @@ public override void VisitProperty(IPropertySymbol symbol)
 
+        private static bool ShouldBeNullableAnnotated(ISymbol symbol, FlowAnalysisAnnotations annotations)
+        {
+            if (symbol is Symbols.PublicModel.Symbol csharpSymbol)
```
>if (symbol is Symbols.PublicModel.Symbol csharpSymbol) [](start = 12, length = 54)

Attributes is not C# specific concept, symbol model allows to check for them even when we are dealing with non-C# symbols. I guess it is fine to use this as an optimization, but it feels wrong to limit the behavior just to C# symbols.

MaStr11 replied:
I rely on the `GetFlowAnalysisAnnotations` extension method which is defined on `Microsoft.CodeAnalysis.CSharp.Symbol`. So this approach relies on CSharp symbols. As I said [earlier](https://github.com/dotnet/roslyn/pull/50527/files#r558384945), I'm not sure about doing the right thing at the right place at all:

Why is `((IPropertySymbol)symbol).Type.NullableAnnotation` not `NullableAnnotation.Annotated` in this case?

AlekseyTs replied:
> Why is ((IPropertySymbol)symbol).Type.NullableAnnotation not NullableAnnotation.Annotated in this case? 

I am not following. The type isn't annotated. Right?

---
In reply to: [558496330](https://github.com/dotnet/roslyn/pull/50527#discussion_r558496330) [](ancestors = 558496330)

MaStr11 replied:
No. the type is not annotated. Another solution would have been to make sure it is annotated and I wasn't sure what the right thing to do is here: Tackle the problem in SymbolVisitor or make sure the type is annotated.

AlekseyTs replied:
>  Another solution would have been to make sure it is annotated and I wasn't sure what the right thing to do is here: Tackle the problem in SymbolVisitor or make sure the type is annotated.

I don't think we would want to change behavior of the ```Type``` on the symbol itself. The annotation exposed there is about an annotation from source, i.e. about the presence or lack of the `?` in source and how the fact is interpreted.

---
In reply to: [558553322](https://github.com/dotnet/roslyn/pull/50527#discussion_r558553322) [](ancestors = 558553322)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,140 @@ public void RecordDeclaration()
+            var format = new SymbolDisplayFormat(
+                memberOptions: SymbolDisplayMemberOptions.IncludeParameters | SymbolDisplayMemberOptions.IncludeType | SymbolDisplayMemberOptions.IncludeModifiers,
+                genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+                miscellaneousOptions: SymbolDisplayMiscellaneousOptions.None); // without SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier
```
>miscellaneousOptions: SymbolDisplayMiscellaneousOptions.None); // without SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier [](start = 16, length = 144)

Is this the same scenario symbol-wise as in the previous unit-test? If so, just add verification with different format options there.

MaStr11 replied:
Done in aaa538f4

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,140 @@ public void RecordDeclaration()
+
+            // look for memberAccess "a.P"
+            var memberAccess = tree.GetRoot().DescendantNodes().OfType<AssignmentExpressionSyntax>().Single().Right;
+            var symbol = model.GetSymbolInfo(memberAccess).Symbol;
```
>var symbol = model.GetSymbolInfo(memberAccess).Symbol; [](start = 12, length = 54)

It is an overkill to use syntax and SemanticModel to get to a symbol for a type memeber. Symbols can be accessed directly from a compilation. We also have helpers like ```GetMember```, etc.

MaStr11 replied:
That pattern is often used in that test class and `GetMember` would return `P` of the open generic type, while I need the closed generic type.

AlekseyTs replied:
> That pattern is often used in that test class and GetMember would return P of the open generic type, while I need the closed generic type.

I explained how to get symbol for the closed type below (for NullableAttributes_MayBeNullOnStructProperty unit-test). We also need to test behavior of the API for an opened type. Another thing to consider is that C# SemanticModel is not going to help with getting VB symbols, whereas getting them through symbol model is going to be pretty much the same.

---
In reply to: [558518218](https://github.com/dotnet/roslyn/pull/50527#discussion_r558518218) [](ancestors = 558518218)

MaStr11 replied:
Done in 71b3eac7351c87d232db82d3a8ed5c62b1f1e29e

MaStr11 replied:
> We also need to test behavior of the API for an opened type

Added verification for open types in ddf3d061 and 13c5cf39

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,140 @@ public void RecordDeclaration()
+    public T P { get; }
+}
+
+public class B
```
It feels like this class is unnecessary.

MaStr11 replied:
Fixed in 4e84165d

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,140 @@ public void RecordDeclaration()
+
+            // look for memberAccess "a.P"
+            var memberAccess = tree.GetRoot().DescendantNodes().OfType<AssignmentExpressionSyntax>().Single().Right;
+            var symbol = model.GetSymbolInfo(memberAccess).Symbol;
```
>var symbol = model.GetSymbolInfo(memberAccess).Symbol; [](start = 12, length = 54)

Same comment.

MaStr11 replied:
Done in 71b3eac7351c87d232db82d3a8ed5c62b1f1e29e

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,140 @@ public void RecordDeclaration()
+            var model = compilation.GetSemanticModel(tree);
+
+            var memberAccess = tree.GetRoot().DescendantNodes().OfType<AssignmentExpressionSyntax>().Single().Right;
+            var symbol = model.GetSymbolInfo(memberAccess).Symbol;
```
>var symbol = model.GetSymbolInfo(memberAccess).Symbol; [](start = 12, length = 54)

For this scenario, I would also avoid using semantic model. There is an API to ```Construct``` a generic type. There is also a trick to get constructed type by using it as base type of some other type, etc.
For example, define class like:
```
class C : A<int> {}
```
Locate the type, get its base type, get member from the base type: ```compilation.GetTypeByMetadataName("C").BaseType.GetMembers("P").Single()```

MaStr11 replied:
That pattern (compilation -> tree -> model -> GetSymbolInfo) is used numerous times in that test class before. I just followed that pattern. With this pattern, it is also clear that I ask for `P` on a closed generic type.

AlekseyTs replied:
> That pattern (compilation -> tree -> model -> GetSymbolInfo) is used numerous times in that test class before. I just followed that pattern. With this pattern, it is also clear that I ask for P on a closed generic type.

I assume that whoever wrote those tests didn't know any better. Usage of  SemanticModel when not necessary creates confusion and also an unnecessary overhead during execution.

---
In reply to: [558516077](https://github.com/dotnet/roslyn/pull/50527#discussion_r558516077) [](ancestors = 558516077)

MaStr11 replied:
Done in 71b3eac7351c87d232db82d3a8ed5c62b1f1e29e

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,140 @@ public void RecordDeclaration()
+
+        [Fact]
+        [WorkItem(48023, "https://github.com/dotnet/roslyn/issues/48023")]
+        public void NullableAttributes_MayBeNullOnClassProperty()
```
>NullableAttributes_MayBeNullOnClassProperty [](start = 20, length = 43)

Please add similar tests with VB symbols.

AlekseyTs commented:
Done with review pass (commit 3)

MaStr11 reviewed (commented):
Thanks for the review. Copuld you also please have a look at the spec for the other nullable attributes: https://github.com/dotnet/roslyn/pull/50527#issuecomment-761051094

MaStr11 commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,171 @@ public void RecordDeclaration()
+        {
+            var sourceVB =
+@"
+Namespace System.Diagnostics.CodeAnalysis
+    <AttributeUsage(AttributeTargets.Field Or AttributeTargets.Parameter Or AttributeTargets.[Property] Or AttributeTargets.ReturnValue)>
+    Public NotInheritable Class MaybeNullAttribute
+        Inherits Attribute
+    End Class
+End Namespace
+
+Namespace N
+    Public Class A(Of T)
+        <System.Diagnostics.CodeAnalysis.MaybeNull>
+        Public ReadOnly Property P As T
+    End Class
+End Namespace
+";
+            var sourceCS =
+@"
+#nullable enable
+
+namespace N
+{
+    public class Test: A<object> { }
+}
+";
+            var compVB = CreateVisualBasicCompilation(GetUniqueName(), sourceVB, referencedAssemblies: new[] { MscorlibRef });
+            var refVB = compVB.EmitToImageReference();
+            var compCS = CreateCompilation(sourceCS, references: new[] { refVB });
+            var member = compCS.GetTypeByMetadataName("N.Test").BaseType().GetMember("P");
+            var format = new SymbolDisplayFormat(
+                memberOptions: SymbolDisplayMemberOptions.IncludeParameters | SymbolDisplayMemberOptions.IncludeType | SymbolDisplayMemberOptions.IncludeModifiers,
+                genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+                miscellaneousOptions: SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier);
+
+            Verify(
+                member.ToDisplayParts(format),
+                "Object P",
```
I would have expected to get `Object? P` here, given that VB defines and applies the expected attribute (given that it doesn't matter where `MaybeNullAttribute` is defined).

MaStr11 replied:
According to the [nullable-metadata](https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-metadata.md) documentation, this scenario (VB defines NRT annotations for C#) is not supported. I think I should therefore simplify the test, to avoid confusion.

MaStr11 replied:
I removed the nullable attributes in the VB code in 60553d24.

MaStr11 commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,185 @@ public void RecordDeclaration()
+    Inherits A(Of Object)
+End Class";
+            var compCS = CreateCompilation(new[] { sourceCS, MaybeNullAttributeDefinition });
+            var refCS = compCS.EmitToImageReference();
```
There is also `ToMetadataReference()`. I'm not sure I picked the right method here, or if it makes sense to test with both kinds of references (the same is true for the next test `NullableAttributes_CSharpReferencesVisualBasic`)

AlekseyTs replied:
>There is also ToMetadataReference(). I'm not sure I picked the right method here, or if it makes sense to test with both kinds of references (the same is true for the next test NullableAttributes_CSharpReferencesVisualBasic)

That one returns a compilation reference. Compilation references cannot be used across language "boundary".

---
In reply to: [559173858](https://github.com/dotnet/roslyn/pull/50527#discussion_r559173858) [](ancestors = 559173858)

MaStr11 commented:
@AlekseyTs I addressed your feedback. I second review would be appreciated. If everything is fine I would add the missing pieces according to my spec https://github.com/dotnet/roslyn/pull/50527#issuecomment-761051094:
* `VisitField`: Annotate type if marked with `[MayBeNull]`
* `VisitParameter`: Annotate parameter type if marked with `[MayBeNull]` or `[MaybeNullWhen]`
* `VisitMethod`: Annotate return type if marked with `[return:MayBeNull]`

MaStr11 commented:
@AlekseyTs I just went on and added the missing pieces as described in https://github.com/dotnet/roslyn/pull/50527#issuecomment-761805287. A second review would be appreciated.

MaStr11 commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -699,7 +725,10 @@ public override void VisitParameter(IParameterSymbol symbol)
                 }
 
-                symbol.Type.Accept(this.NotFirstVisitor);
+                var symbolType = ShouldBeNullableAnnotated(symbol, symbol.Type, FlowAnalysisAnnotations.MaybeNull)
```
```suggestion
                var symbolType = ShouldBeNullableAnnotated(symbol, symbol.Type, FlowAnalysisAnnotations.MaybeNullWhenTrue | FlowAnalysisAnnotations.MaybeNullWhenFalse | FlowAnalysisAnnotations.MaybeNull)
```
Parameters are annoted, if `[MaybeNull]`, `[MaybeNullWhen(true)]` or `[MaybeNullWhen(false)]` are present. To make this more obvious I could pass 
`FlowAnalysisAnnotations.MaybeNullWhenTrue | FlowAnalysisAnnotations.MaybeNullWhenFalse | FlowAnalysisAnnotations.MaybeNull`, but `FlowAnalysisAnnotations.MaybeNull` is defined as `MaybeNull = MaybeNullWhenTrue | MaybeNullWhenFalse,`, so it's technically not necessary.

Should I be explicit and add `MaybeNullWhenTrue` and `MaybeNullWhenFalse` to be more coherent?

https://github.com/dotnet/roslyn/blob/f5f54f773cedf07e5c646429be9adad520b401e3/src/Compilers/CSharp/Portable/Symbols/FlowAnalysisAnnotations.cs#L11-L26

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,367 @@ public void RecordDeclaration()
+End Class";
+            var compCS = CreateCompilation(new[] { sourceCS, MaybeNullAttributeDefinition });
+            var refCS = compCS.EmitToImageReference();
+            var compVB = CreateVisualBasicCompilation(GetUniqueName(), sourceVB, referencedAssemblies: new[] { MscorlibRef, refCS });
```
>MscorlibRef [](start = 111, length = 11)

Why do we need to pass this reference explicitly? This is unexpected and is likely to cause issues with the test in the future. If there is some kind of reference mismatch between the compilations, I think we should ensure they use the same target framework instead.

MaStr11 replied:
I replaced it with `TargetFrameworkUtil.GetReferences(TargetFramework.Standard, new[] { refCS })` in b6e79a8b. If I just drop MscorlibRef, `object` can't be resolved.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7652,5 +7652,367 @@ public void RecordDeclaration()
+
+            Verify(
+                closedGenericMember.ToDisplayParts(format),
+                "Overloads P As Object",
```
>"Overloads P As Object", [](start = 16, length = 24)

This isn't quite what I had in mind. It looks like we are testing VB's SymbolDisplay, which this PR doesn't change. Even though it is good to have a test like this for completeness, it is more interesting to test behavior of C#'s SymbolDisplay when it is fed VB symbols.

MaStr11 replied:
I added `NullableAttributes_VisualBasicReferencesCSharp` (this test) and `NullableAttributes_CSharpReferencesVisualBasic`, so there is a test for the other way around.

AlekseyTs replied:
> I added NullableAttributes_VisualBasicReferencesCSharp (this test) and NullableAttributes_CSharpReferencesVisualBasic, so there is a test for the other way around.

The ```NullableAttributes_CSharpReferencesVisualBasic``` tests something else. It looks like it tests C# SymbolDisplay for C# symbols representing members originally declared in VB. In my opinion, this isn't particularly interesting since VB declarations in the test don't do anything that cannot be done with declarations in C#. The test that we are missing is C# SymbolDisplay for VB symbol in this test. See how ```TestMethodVB``` in this file achieves that. We want to test changes that you are making in SymbolDisplayVisitor.Members.cs with VB symbols as an input.

---
In reply to: [561243985](https://github.com/dotnet/roslyn/pull/50527#discussion_r561243985) [](ancestors = 561243985)

MaStr11 replied:
Sorry for getting this wrong and thanks for the pointer to a sample test. I added a new test in 569d0ea7 that should test the right thing now. One kind of symbol, that is not covered yet are `SignatureOnly...` symbols. Azure pipelines CI revealed that (see [test-results](https://dev.azure.com/dnceng/public/_build/results?buildId=957883&view=ms.vss-test-web.build-test-results-tab&runId=30203184&resultId=107662&paneView=debug)). I fixed that in 388097d. Do we need tests for this kind of symbol too?

AlekseyTs replied:
> I added a new test in 569d0ea7 that should test the right thing now.

NullableAttributes_SymbolDisplayForVBSymbol is good to have. However, I would like us to have VB symbols from **this test** pushed through C# SymbolDisplay. It is interesting to see how the same members declared in C# displayed differently or not differently when looked at through VB symbol model. Sorry for not making this clear from the beginning.

---
In reply to: [561753871](https://github.com/dotnet/roslyn/pull/50527#discussion_r561753871) [](ancestors = 561753871)

AlekseyTs replied:
> One kind of symbol, that is not covered yet are ```SignatureOnly...``` symbols. 

I am somewhat surprised that ```SignatureOnly...``` symbols are leaked through diagnostics. Perhaps this (the leak) is something we should take a closer look at. Could you find out what specific error does that in the identified test? Since now we a "losing" control over usage of ```FlowAnalysisAnnotations``` API, I think we need to make sure that it doesn't throw without a good reason. Please review all implementations of this API and make sure we no longer assume that the code is unreachable. Then, I would be comfortable with not having targeted SymbolDisplay tests for all different implementations of the API. 

---
In reply to: [561978351](https://github.com/dotnet/roslyn/pull/50527#discussion_r561978351) [](ancestors = 561978351,561753871)

MaStr11 replied:
I see. I think f122b80 is what you were asking for.

MaStr11 replied:
> I am somewhat surprised that SignatureOnly... symbols are leaked through diagnostics. Could you find out what specific error does that in the identified test?

`SignatureOnlyParameterSymbol` is used to improve the diagnostic message in overload resolution (note: the list all classes wich construct SignatureOnlyParameterSymbol, but not all places, I have not looked for other `SignatureOnly..` symbols):

https://github.com/dotnet/roslyn/blob/ff498f2fd470b6ff2b956ac0c4317667281688ad/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs#L1260-L1282

There are also other places where `SignatureOnlyParameterSymbol` is created, that do not seem to be related to diagnostics:

Binder_Crefs
https://github.com/dotnet/roslyn/blob/ff498f2fd470b6ff2b956ac0c4317667281688ad/src/Compilers/CSharp/Portable/Binder/Binder_Crefs.cs#L885-L900

`dynamic` handling in Conversions.cs
https://github.com/dotnet/roslyn/blob/ff498f2fd470b6ff2b956ac0c4317667281688ad/src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs#L250-L272

Retargeting:
https://github.com/dotnet/roslyn/blob/ff498f2fd470b6ff2b956ac0c4317667281688ad/src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.cs#L1030-L1042

Records member synthesizing:
https://github.com/dotnet/roslyn/blob/ff498f2fd470b6ff2b956ac0c4317667281688ad/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L3140-L3157

MaStr11 replied:
> Please review all implementations of this API and make sure we no longer assume that the code is unreachable. Then, I would be comfortable with not having targeted SymbolDisplay tests for all different implementations of the API.

There are 8 types, that override the abstract `ParameterSymbol.FlowAnalysisAnnotations` property. `SymbolOnlyParameterSymbol` was the only one, that was throwing. All others either return `None` (FunctionPointerParameterSymbol, SignatureOnlyParameterSymbol, SourceClonedParameterSymbol, SourceSimpleParameterSymbol, SynthesizedParameterSymbolBase, ThisParameterSymbol), delegate to the underlying symbol (WrappedParameterSymbol) or decode the information (SourceComplexParameterSymbol).

The property itself is exclusively used directly in the NullableWalker (and in `SourceMemberContainerTypeSymbol.CheckValidNullableMethodOverride`). All other accesses are through `SymbolExtensions.GetFlowAnalysisAnnotations`.  I would not expect `SymbolExtensions.GetFlowAnalysisAnnotations` to throw in any case.

I only analyzed ParameterSymbol. If you like, I good do the same for the other affected symbols.

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -172,6 +178,23 @@ public override void VisitProperty(IPropertySymbol symbol)
+            {
+                var flowAnalysisAnnotations = csharpSymbol.UnderlyingSymbol.GetFlowAnalysisAnnotations();
+                // Is any flag in flowAnalysisAnnotations also present in annotations?
+                return (annotations & flowAnalysisAnnotations) > 0;
```
>> [](start = 63, length = 1)

When we test bits, we usually use ```==```/```!=```

MaStr11 replied:
Done in fd3b6578

AlekseyTs commented:
Done with review pass (commit 18)

AlekseyTs commented:
@MaStr11 It looks like there are legitimate test failures.

AlekseyTs commented:
@CyrusNajmabadi, @jcouv What do you think about proposed behavior changes in SymbolDisplay around nullable attributes? Should we add a special display option to control the behavior?

MaStr11 reviewed (commented):
> It looks like there are legitimate test failures.

Yes. The failing tests are caused by the new `?` in error messages related to the placement of the nullable attributes (It shouldn't be a big deal to fix, but it will bloat the PR). I will wait with resolving these until the sign off from @CyrusNajmabadi and @jcouv https://github.com/dotnet/roslyn/pull/50527#issuecomment-763868378.

AlekseyTs commented:
> I will wait with resolving these until the sign off from

In order to form an opinion about the impact of a change it is often very useful to be able to see how exactly the change affects behavior of existing tests (through baseline changes, etc.). For example, for this particular change this might help reviewers to decide if we actually need a special display option.

AlekseyTs commented:
Done with review pass (commit 20)

AlekseyTs commented:
@CyrusNajmabadi, @jcouv  Ping. It would be interesting to hear what you think. I am on the fence about translating attributes to question marks by default, for example. The way it affects diagnostics, etc.

---
In reply to: [763868378](https://github.com/dotnet/roslyn/pull/50527#issuecomment-763868378) [](ancestors = 763868378)

AlekseyTs commented:
Done with review pass (commit 35)

jcouv commented:
Thanks for the ping.
I'm curious to hear what @CyrusNajmabadi thinks, but I'm rather opposed to this change. 
I feel we should not fudge symbols and how they are displayed, because it's not accurately describing what is there (causes confusion) and the approach will fall short quickly (we can only solve a simple case with this trick, but conditional attributes and one-way restrictions are not amenable to such representation).

As an alternative:
1. the IDE could offer a fix whenever a nullable attribute could be replaced by a `?` annotation (especially with the C# 9 extension on `T?`). This seems a more robust way of reducing usage of attributes wherever possible.
2. maybe QuickInfo should include nullability attributes?

MaStr11 commented:
> maybe QuickInfo should include nullability attributes?

I do back this idea and proposed it in https://github.com/dotnet/roslyn/pull/50527#issuecomment-761051094 as an alternative solution.

CyrusNajmabadi commented:
Oh, this is a very interesting question.  In general i mostly agree with julien here: https://github.com/dotnet/roslyn/pull/50527#issuecomment-764982542

However, it does somewhat come down to a question around how the lang/compiler thinks of these symbols.  For example, if i have:

```c#
public class A<T>
{
    [return: MaybeNull]
    public T M() => default;
}
```

Is that effectively equivalent to:

```c#
public class A<T>
{
    public T? M() => default;
}
```

In other words, is the new form just a more convenient way of stating the older attribute-based form?  Or are there distinct differences in how either would be treated by the compiler later on during NRT analysis?

If they are effectively the same, the i support having the symbol model effectively model that.  But that would include both things like SYmbolDisplay, as well as just the core methods on ITypeSymbol (and the rest) as well.  

If they are not hte same, then i don't think we should do this.  And i instead think we should likely have IDE features like Julien mentioned to both push people to the 'more modern' form, as well as provide this information in our language services to help users.

RikkiGibson commented:
I thought maybe this discussion would be useful: https://github.com/dotnet/csharplang/discussions/4102#discussioncomment-128346. It's specific to parameters but can be extrapolated to return types.

Particularly:

> we found that in the absence of T?, developers used [AllowNull] a stand-in for T? on input parameters.
>
> It feels like the concept of a separate "input state" and "output state" on input parameters is not desirable to include by default.
>
> Instead, we think the better mental model is to say that an input parameter usually only has an input state.

What I'd draw from this is:
- If the parameter is not `ref`/`out` then `[AllowNull] T` is equivalent to `T?`.
- If a return value is not `ref` then `[MaybeNull] T` is equivalent to `T?`.
- An `out` parameter like `[MaybeNull] out T result` is equivalent to `out T? result`. This doesn't apply for `ref` parameters.
- This should still hold when certain combinations of multiple attributes are used, i.e. `void MyAssert([NotNull, AllowNull] T param)` is expected to be equivalent to `void MyAssert([NotNull] T? param)`. I wouldn't be surprised to find bugs/edge cases here though.

I think if we ended up with tooling that suggests to convert the `AllowNull/NotNull` to `T?`, for example, in the first 3 cases above, that would be great.

I could be over-generalizing here, so please check me if this doesn't seem right to you @jcouv @333fred.

RikkiGibson commented:
Also wanted to elaborate on the `ref` scenario. A `ref` parameter or return is distinct by having both an "input" state and "output" state which is controlled by `AllowNull/DisallowNull` and `MaybeNull/NotNull` respectively. So a simplification is likely possible if for example a `ref` parameter has "corresponding" input and output attributes, for example `[AllowNull, MaybeNull] ref T t` is equivalent to `ref T? t`.

333fred commented:
I think I would likely prefer to go the route of tooling suggesting moving towards `?`, instead of trying to fudge the symbols. Fudging symbols tends to lead to the types of rabbit holes of "I need to cover these 500 scenarios that we didn't consider when agreeing to do the thing".

RikkiGibson commented:
I agree that we should avoid trying to convert nullability attributes to `?` annotations in symbol display. I could imagine cases where adding the nullability attributes to the symbol display could be helpful, though.

333fred commented:
I might not be opposed to adding a SymbolDisplayKind option to explicitly turn these attributes into annotations, which the IDE could use if they wanted to. @AlekseyTs, what do you think of such an option?

jcouv commented:
@CyrusNajmabadi `[MaybeNull]T` and `T?` behave the same certain in positions.
But as far as compiler APIs go, they are different, one has a `[MaybeNull]` attribute and the other will end up having a synthesized `[Nullable]` attribute of some sort. So the compiler APIs (annotations, attributes) should behave differently on both of those. That precludes option 4 below (changing symbols more broadly).
I'd still recommend option 1 (IDE fixer instead).

Options:
1. no compiler API change, help users migrate to more modern `T?` in source instead
2. only tweak symbol APIs for display
3. add a new symbol display option
4. adjust symbol APIs more broadly (display, annotations, attributes)

333fred commented:
I'd go 1 or 3.

AlekseyTs commented:
> I might not be opposed to adding a SymbolDisplayKind option to explicitly turn these attributes into annotations, which the IDE could use if they wanted to. @AlekseyTs, what do you think of such an option?

I prefer adding a special display option rather than changing default behavior

AlekseyTs commented:
I do not have strong opinion about whether behavior of SymbolDispalay should change at all. But it feels somewhat strange to me to translate attributes to syntax that doesn't exist in source. Could be confusing to consumers.

CyrusNajmabadi commented:
I would do 1 or 4, with strong leaning toward 1.

jcouv commented:
Alright then. I'll go ahead and close this PR (and the bug that prompted it).
@MaStr11 Sorry for the trouble. 

I'd suggest we start with a fixer to convert `[MaybeNull]T` (on outputs) and `[AllowNull]T` (on inputs) to `T?` first. 
Filed https://github.com/dotnet/roslyn/issues/50819

Separately, the IDE team should consider whether to display nullability attributes. 
We already have issue https://github.com/dotnet/roslyn/issues/38358 tracking that.
